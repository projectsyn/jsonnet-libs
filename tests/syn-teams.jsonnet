local com = import 'lib/commodore.libjsonnet';
local syn_teams = import 'syn-teams.libsonnet';

local expected = com.inventory().expected;

local teams =
  local t = std.set(syn_teams.teams());
  local expected_teams = std.set(expected.teams);
  if t != expected_teams then
    error 'Expected teams: %s, got teams: %s' % [
      std.join(', ', teams),
      std.join(', ', expected_teams),
    ]
  else
    t;

local all_teams =
  local t = std.set(syn_teams.teams(true));
  local expected_teams = std.set(expected.all_teams);
  if t != expected_teams then
    error 'Expected teams: %s, got teams: %s' % [
      std.join(', ', teams),
      std.join(', ', expected_teams),
    ]
  else
    t;

local appsForTeam =
  local aft = {
    [t]: syn_teams.applicationsForTeam(t)
    for t in all_teams
  };
  if aft != expected.appsForTeam then
    error 'Mismatch in appsForTeam: got %s, expected %s' % [
      aft,
      expected.appsForTeam,
    ]
  else
    aft;

local teamForApp =
  local inv = com.inventory();
  local applications = std.map(function(app) syn_teams.appKeys(app, true)[0], inv.applications);
  local tfa = {
    [a]: syn_teams.teamForApplication(a)
    for a in applications
  };
  if tfa != expected.teamForApp then
    error 'Mismatch in teamForApp: got %s, expected %s' % [
      tfa,
      expected.teamForApp,
    ]
  else
    tfa;

local appKeys =
  local expected = {
    foo: [ 'foo' ],
    'foo-bar': [ 'foo_bar' ],
    'foo-bar as bar': [ 'bar', 'foo_bar' ],
    'foo-bar as bar-qux': [ 'bar_qux', 'foo_bar' ],
  };

  {
    local ks = syn_teams.appKeys(a),
    [a]: if ks != expected[a] then
      error 'Expected %s for appKeys(%s), got %s' % [
        expected[a],
        a,
        ks,
      ]
    else
      ks
    for a in std.objectFields(expected)
  };

local appKeysRaw =
  local expected = {
    foo: [ 'foo' ],
    'foo-bar': [ 'foo-bar' ],
    'foo-bar as bar': [ 'bar', 'foo-bar' ],
    'foo-bar as bar-qux': [ 'bar-qux', 'foo-bar' ],
  };

  {
    local ks = syn_teams.appKeys(a, true),
    [a]: if ks != expected[a] then
      error 'Expected %s for appKeys(%s), got %s' % [
        expected[a],
        a,
        ks,
      ]
    else
      ks
    for a in std.objectFields(expected)
  };


{
  teams: teams,
  all_teams: all_teams,
  appKeys: appKeys,
  appKeysRaw: appKeysRaw,
  appsForTeam: appsForTeam,
  teamForApp: teamForApp,
}
