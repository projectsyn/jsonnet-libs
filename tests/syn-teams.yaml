applications:
  - small-sun
  - cool-breeze
  - cool-breeze as red-flower
  - fragrant-smoke

parameters:
  syn:
    owner: sparkling-sound
    teams:
      solitary-wood:
        instances:
          - small-sun
      lingering-fog:
        instances:
          - morning-thunder
      fragrant-flower:
        instances:
          - ~small-sun
          - red-flower


expected:
  teams:
    - solitary-wood
    - fragrant-flower
  all_teams:
    - sparkling-sound
    - solitary-wood
    - fragrant-flower
  appsForTeam:
    fragrant-flower:
      - red-flower
    sparkling-sound:
      - cool-breeze
      - fragrant-smoke
    solitary-wood:
      - small-sun
  teamForApp:
    cool-breeze: sparkling-sound
    fragrant-smoke: sparkling-sound
    red-flower: fragrant-flower
    small-sun: solitary-wood
