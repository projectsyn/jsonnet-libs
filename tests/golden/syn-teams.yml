{
   "all_teams": [
      "fragrant-flower",
      "solitary-wood",
      "sparkling-sound"
   ],
   "appKeys": {
      "foo": [
         "foo"
      ],
      "foo-bar": [
         "foo_bar"
      ],
      "foo-bar as bar": [
         "bar",
         "foo_bar"
      ],
      "foo-bar as bar-qux": [
         "bar_qux",
         "foo_bar"
      ]
   },
   "appKeysRaw": {
      "foo": [
         "foo"
      ],
      "foo-bar": [
         "foo-bar"
      ],
      "foo-bar as bar": [
         "bar",
         "foo-bar"
      ],
      "foo-bar as bar-qux": [
         "bar-qux",
         "foo-bar"
      ]
   },
   "appsForTeam": {
      "fragrant-flower": [
         "red-flower"
      ],
      "solitary-wood": [
         "small-sun"
      ],
      "sparkling-sound": [
         "cool-breeze",
         "fragrant-smoke"
      ]
   },
   "teamForApp": {
      "cool-breeze": "sparkling-sound",
      "fragrant-smoke": "sparkling-sound",
      "red-flower": "fragrant-flower",
      "small-sun": "solitary-wood"
   },
   "teams": [
      "fragrant-flower",
      "solitary-wood"
   ]
}
